compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs

```diff
@@ -233,6 +238,50 @@ pub fn push_debuginfo_type_name<'tcx>(
         }
     }
 
+    fn msvc_enum_fallback(
+        tcx: TyCtxt<'tcx>,
+        ty: Ty<'tcx>,
+        def: &AdtDef,
+        substs: SubstsRef<'tcx>,
+        output: &mut String,
+        visited: &mut FxHashSet<Ty<'tcx>>,
+    ) {
+        let layout = tcx.layout_of(tcx.param_env(def.did).and(ty)).expect("layout error");
+
+        if let Variants::Multiple {
+            tag_encoding: TagEncoding::Niche { dataful_variant, .. },
+            tag,
+            variants,
+            ..
+        } = &layout.variants
+        {
+            let dataful_variant_layout = &variants[*dataful_variant];
+
+            // calculate the range of values for the dataful variant
+            let dataful_discriminant_range =
+                &dataful_variant_layout.largest_niche.as_ref().unwrap().scalar.valid_range;
+
+            let min = dataful_discriminant_range.start();
+            let min = tag.value.size(&tcx).truncate(*min);
+
+            let max = dataful_discriminant_range.end();
+            let max = tag.value.size(&tcx).truncate(*max);
+
+            output.push_str("_enum<");
```

Comment:
Would it be possible to use a name that is not a valid Rust type name?  For example, "$enum", or "$rust::enum".

(And, since this is already a breaking change, maybe do the same for "tuple" and "slice"?)

Reply:
I think this is a good idea but I would like to leave this for a follow-up PR. Some general guidelines for msvc debuginfo are being drafted in #85269 and this would be a good one to add there. The reason I used `_enum` here instead of `enum` is that the raw identifier `enum` seems to be treated as some kind of keyword by the DIA parser and no natvis will match on it.

Reply:
I'll make sure to update this once we decide on a consistent format - if we were going with a name that works in MSVC's debugger, but not Rust it would use a trailing `$`: `enum$`.

Reply:
I went ahead and changed this to be inline with the consensus that's forming in #85269. Thanks for the suggestion @vadimcn!